File: app/src/main/java/com/micewine/emu/views/JoystickView.java
package com.micewine.emu.views;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

public class JoystickView extends SurfaceView implements SurfaceHolder.Callback {
    private final int JOYSTICK_RADIUS = 100;
    private final int BUTTON_RADIUS = 50;
    private final int DPAD_SIZE = 150;

    private float joystickCenterX, joystickCenterY;
    private boolean isJoystickPressed = false;

    private float button1CenterX, button1CenterY;
    private boolean isButton1Pressed = false;

    private float button2CenterX, button2CenterY;
    private boolean isButton2Pressed = false;

    private Paint paint;

    public JoystickView(Context context) {
        super(context);
        init();
    }

    public JoystickView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public JoystickView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        getHolder().addCallback(this);
        paint = new Paint();
        paint.setAntiAlias(true);
        paint.setColor(Color.BLUE);
    }

    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        // Initialize control positions
        joystickCenterX = getWidth() / 2;
        joystickCenterY = getHeight() / 2;

        button1CenterX = getWidth() / 4;
        button1CenterY = getHeight() / 4;

        button2CenterX = getWidth() * 3 / 4;
        button2CenterY = getHeight() / 4;

        drawControls();
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
        // Not needed
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {
        // Not needed
    }

    private void drawControls() {
        Canvas canvas = getHolder().lockCanvas();
        if (canvas != null) {
            canvas.drawColor(Color.WHITE);
            // Draw joystick
            paint.setColor(Color.BLUE);
            canvas.drawCircle(joystickCenterX, joystickCenterY, JOYSTICK_RADIUS, paint);
            // Draw button 1
            paint.setColor(isButton1Pressed ? Color.RED : Color.GREEN);
            canvas.drawCircle(button1CenterX, button1CenterY, BUTTON_RADIUS, paint);
            // Draw button 2
            paint.setColor(isButton2Pressed ? Color.RED : Color.GREEN);
            canvas.drawCircle(button2CenterX, button2CenterY, BUTTON_RADIUS, paint);
            // Draw D-pad
            paint.setColor(Color.BLACK);
            canvas.drawRect(0, getHeight() - DPAD_SIZE, DPAD_SIZE, getHeight(), paint);
            getHolder().unlockCanvasAndPost(canvas);
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
            case MotionEvent.ACTION_MOVE:
                float x = event.getX();
                float y = event.getY();

                // Joystick
                float dx = x - joystickCenterX;
                float dy = y - joystickCenterY;
                if (Math.sqrt(dx * dx + dy * dy) <= JOYSTICK_RADIUS) {
                    joystickCenterX = x;
                    joystickCenterY = y;
                    isJoystickPressed = true;
                } else {
                    isJoystickPressed = false;
                }

                // Button 1
                if (Math.sqrt(Math.pow(x - button1CenterX, 2) + Math.pow(y - button1CenterY, 2)) <= BUTTON_RADIUS) {
                    isButton1Pressed = true;
                } else {
                    isButton1Pressed = false;
                }

                // Button 2
                if (Math.sqrt(Math.pow(x - button2CenterX, 2) + Math.pow(y - button2CenterY, 2)) <= BUTTON_RADIUS) {
                    isButton2Pressed = true;
                } else {
                    isButton2Pressed = false;
                }

                drawControls();
                break;
            case MotionEvent.ACTION_UP:
                isJoystickPressed = false;
                isButton1Pressed = false;
                isButton2Pressed = false;
                drawControls();
                break;
        }
        return true;
    }
}

File: app/src/main/java/com/micewine/emu/views/JoystickView.kt
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.SurfaceHolder
import android.view.SurfaceView

class JoystickView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : SurfaceView(context, attrs, defStyleAttr), SurfaceHolder.Callback {
    private val joystickRadius = 100f
    private val buttonRadius = 50f
    private val dpadSize = 150

    private var joystickCenterX: Float = 0f
    private var joystickCenterY: Float = 0f
    private var isJoystickPressed = false

    private var button1CenterX: Float = 0f
    private var button1CenterY: Float = 0f
    private var isButton1Pressed = false

    private var button2CenterX: Float = 0f
    private var button2CenterY: Float = 0f
    private var isButton2Pressed = false

    private var paint: Paint = Paint()

    init {
        holder.addCallback(this)
        paint.isAntiAlias = true
        paint.color = Color.BLUE
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        // Initialize control positions
        joystickCenterX = width / 2f
        joystickCenterY = height / 2f

        button1CenterX = width / 4f
        button1CenterY = height / 4f

        button2CenterX = width * 3f / 4f
        button2CenterY = height / 4f

        drawControls()
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        // Not needed
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        // Not needed
    }

    private fun drawControls() {
        val canvas = holder.lockCanvas()
        canvas?.let {
            it.drawColor(Color.WHITE)
            // Draw joystick
            paint.color = Color.BLUE
            it.drawCircle(joystickCenterX, joystickCenterY, joystickRadius, paint)
            // Draw button 1
            paint.color = if (isButton1Pressed) Color.RED else Color.GREEN
            it.drawCircle(button1CenterX, button1CenterY, buttonRadius, paint)
            // Draw button 2
            paint.color = if (isButton2Pressed) Color.RED else Color.GREEN
            it.drawCircle(button2CenterX, button2CenterY, buttonRadius, paint)
            // Draw D-pad
            paint.color = Color.BLACK
            it.drawRect(0f, (height - dpadSize).toFloat(), dpadSize.toFloat(), height.toFloat(), paint)
            holder.unlockCanvasAndPost(it)
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN, MotionEvent.ACTION_MOVE -> {
                val x = event.x
                val y = event.y

                // Joystick
                val dx = x - joystickCenterX
                val dy = y - joystickCenterY
                isJoystickPressed = Math.sqrt((dx * dx + dy * dy).toDouble()) <= joystickRadius
                if (isJoystickPressed) {
                    joystickCenterX = x
                    joystickCenterY = y
                }

                // Button 1
                isButton1Pressed = Math.sqrt(((x - button1CenterX) * (x - button1CenterX) + (y - button1CenterY) * (y - button1CenterY)).toDouble()) <= buttonRadius

                // Button 2
                isButton2Pressed = Math.sqrt(((x - button2CenterX) * (x - button2CenterX) + (y - button2CenterY) * (y - button2CenterY)).toDouble()) <= buttonRadius

                drawControls()
            }
            MotionEvent.ACTION_UP -> {
                isJoystickPressed = false
                isButton1Pressed = false
                isButton2Pressed = false
                drawControls()
            }
        }
        return true
    }
}
